apps/state/lib/state/server/query.ex

```diff
@@ -0,0 +1,182 @@
+defmodule State.Server.Query do
+  @moduledoc """
+  Queries a Server given a map of values.
+
+  Given a Server module and a map of key -> [value], return the values from
+  the Server where each of the keys has one of the values from the given
+  list. A key without a provided list can have any value.
+
+  An example query for schedules would look like:
+
+  %{
+     trip_id: ["1", "2"],
+     stop_id: ["3", "4"]
+  }
+
+  And the results would be the schedules on trip 1 or 2, stopping at stop 3 or 4.
+  """
+
+  @type q :: map
+  @type recordable :: struct
+  @type index :: atom
+
+  alias State.Server
+
+  @spec query(module, q | [q]) :: [recordable] when q: map, recordable: struct
+  def query(module, %{} = q) when is_atom(module) do
+    do_query(module, [q])
+  end
+
+  def query(module, [%{} | _] = qs) do
+    do_query(module, qs)
+  end
+
+  def query(_module, []) do
+    []
+  end
+
+  defmodule Result do
+    @moduledoc false
+    @enforce_keys [:module]
+    defstruct [:module, matches: [], selectors: []]
+  end
+
+  alias __MODULE__.Result
+
+  defp do_query(module, qs) do
+    qs
+    |> Enum.reduce(%Result{module: module}, &accumulate/2)
+    |> finalize()
+  end
+
+  def accumulate(q, %Result{module: module} = result) when map_size(q) > 0 do
+    {is_db_index?, index} = first_index(module.indices(), q)
+    rest = Map.delete(q, index)
+    recordable = module.recordable()
+
+    case {is_db_index?,
+          Enum.reduce_while(rest, {recordable.filled(:_), [], 1}, &build_struct_and_guards/2)} do
+      {true, {struct, [], _}} ->
+        index_values = Map.get(q, index)
+
+        matches =
+          for record <- records_from_struct_and_values(struct, index, index_values) do
+            Server.by_index_match(record, module, {index, module.key_index()}, [])
+          end
+
+        %{result | matches: matches ++ result.matches}
+
+      {_, {struct, guards, _}} ->
+        # put shorter guards at the front
+        guards = Enum.sort(guards)
+
+        index_values = Map.get(q, index)
+
+        selectors =
+          for record <- records_from_struct_and_values(struct, index, index_values) do
+            {
+              record,
+              guards,
+              [:"$_"]
+            }
+          end
+
+        %{result | selectors: [selectors | result.selectors]}
```

Comment:
Should this be `[selectors ++ result.selectors]`?

Reply:
No, they get flattened in `finalize`:

```elixir
  def finalize(%Result{module: module, selectors: selectors, matches: matches}) do
    selected = Server.select_with_selectors(module, List.flatten(selectors))
    flat_results([selected | matches])
  end
```